hoomd/RandomNumbers.h

```diff
@@ -0,0 +1,513 @@
+        */
+        template<typename RNG>
+        DEVICE inline Real operator()(RNG& rng)
+            {
+            return a + (b-a)*generate_canonical<Real>(rng);
+            }
+
+    private:
+        const Real a;     //!< Left end point of the interval
+        const Real b;     //!< Right end point of the interval
+    };
+
+//! Generate a normally distributed random value
+/*! Use the Box-Muller method to generate normally distributed random values.
+*/
+template<typename Real>
+class NormalDistribution
+    {
+    public:
+        //! Constructor
+        /*! \param _sigma Standard deviation of the distribution
+            \param _mu Mean of the distribution
+        */
+        DEVICE explicit NormalDistribution(Real _sigma=Real(1.0), Real _mu=Real(0.0))
+            : sigma(_sigma), mu(_mu)
+            {
+            }
+
+        //! Draw a value from the distribution
+        /*! \param rng Random number generator
+            \returns normally distributed random value in with standard deviation *sigma* and mean *mu*.
+        */
+        template<typename RNG>
+        DEVICE inline Real operator()(RNG& rng)
+            {
+            uint64_t u0, u1;
+            generate_2u64(u0, u1, rng);
+
+            // from random123/examples/boxmuller.hpp
+            Real x, y;
+            fast::sincospi(r123::uneg11<float>(u0), x, y);
+            Real r = sqrtf(-2.f * logf(r123::u01<float>(u1))); // u01 is guaranteed to avoid 0.
+            x *= r;
+            y *= r;
+            return x * sigma + mu;
+            }
+
+        //! Draw two values from the distribution
+        /*! \param out1 [out] First output
+            \param out2 [out] Second output
+            \param rng Random number generator
+            \returns normally distributed random value in with standard deviation *sigma* and mean *mu*.
+        */
+        template<typename RNG>
+        DEVICE inline void operator()(Real& out1, Real& out2, RNG& rng)
+            {
+            uint64_t u0, u1;
+            generate_2u64(u0, u1, rng);
+
+            // from random123/examples/boxmuller.hpp
+            Real x, y;
+            fast::sincospi(r123::uneg11<float>(u0), x, y);
+            Real r = sqrtf(-2.f * logf(r123::u01<float>(u1))); // u01 is guaranteed to avoid 0.
+            x *= r;
+            y *= r;
+            out1 = x * sigma + mu;
+            out2 = y * sigma + mu;
+            }
+
+    private:
+        const Real sigma;     //!< Standard deviation
+        const Real mu;        //!< Mean
+    };
+
+//! Generate a random point on the surface of a sphere
+template<typename Real>
+class SpherePointGenerator
+    {
+    public:
+        DEVICE explicit SpherePointGenerator() {}
+
+        template<typename GeneratorType, typename Real3>
+        DEVICE inline void operator()(GeneratorType& rng, Real3& point)
+            {
+            // draw a random angle
+            const Real theta = rng.s(Real(0), Real(MPCD_2PI));
```

Comment:
`theta` should be drawn using the UniformGenerator. And then Saru can be replaced by the RandomGenerator in the MPCD code. Let me know if you want me to handle these changes though, since I am more familiar with that code.